sycl/doc/extensions/proposed/sycl_ext_oneapi_peer_access.asciidoc

```
@@ -0,0 +1,162 @@
+                                  ext::oneapi::peer_access value =
+                                    ext::oneapi::peer_access::access_supported);
+  void ext_oneapi_enable_peer_access(const device &peer);
+  void ext_oneapi_disable_peer_access(const device &peer);
+};
+
+} // namespace sycl
+----
+
+The semantics of the new functions are:
+
+|===
+|Member Function |Description
+
+|bool ext_oneapi_can_access_peer(const device &peer,
+                                  ext::oneapi::peer_access value =
+                                  ext::oneapi::peer_access::access_supported)
+a|Queries the peer access status between this device and `peer` according to
+the query `value`:
+
+* `ext::oneapi::peer_access::access_supported`: Returns true only if it is
+possible for this device to enable peer access to USM device memory allocations
+located on the `peer` device.
+
+* `ext::oneapi::peer_access::access_enabled`: Returns true only if peer access is
+currently enabled from this device to the `peer` device.
+
+* `ext::oneapi::peer_access::atomics_supported`: When this query returns true,
+it indicates that this device may concurrently perform atomic operations on USM
+device memory allocations located on the `peer` device when peer access is enabled
+to that device. Atomics performed on a peer device's memory must have
+`memory_scope::system` scope.
+If the query returns false, attempting to perform atomic operations
+on `peer` memory will have undefined behavior.
+
+|void enable_peer_access(const device &peer)
+|Enables this device to access USM device allocations located on the peer
+device. This does not permit the peer device to access this device's memory.
+Once this access is enabled, SYCL kernel functions and the explicit memory
+functions may access USM device allocations on the peer device subject to the
+normal rules about context as described in the core SYCL specification.
+Throws an exception if access cannot be enabled or if access is already
+enabled.
```

This should clarify which exception is thrown.  Something like:

> If this device does not support peer access (as defined by `peer_access::access_supported`), throws an `exceptions` with the `errc::feature_not_supported` error code.  If access is already enabled, throws an `exception` with the `errc::invalid` error code.

Can you refresh my memory, though, about why we want to throw an exception if access is already supported?  Couldn't the function just do nothing in that case?

---

Currently this happens in this case in the draft impl:

```
PI CUDA ERROR:
	Value:           704
	Name:            CUDA_ERROR_PEER_ACCESS_ALREADY_ENABLED
	Description:     peer access is already enabled
	Function:        cuda_piextEnablePeer
	Source Location: llvm/sycl/plugins/cuda/pi_cuda.cpp:5625
```